File: XRViewer/Overlay/UIOverlayController.m
#import "UIOverlayController.h"
#import "OverlayViewController.h"
#import "XRViewer-Swift.h"

@interface UIOverlayController ()

@property (nonatomic, weak) UIView *rootView;
@property (nonatomic, strong) TouchView *touchView;
@property (nonatomic, strong) UIWindow *overlayWindow;
@property (nonatomic, strong) OverlayViewController *overlayVC;

@property (nonatomic, copy) HotAction cameraAction;
@property (nonatomic, copy) HotAction micAction;
@property (nonatomic, copy) HotAction showAction;
@property (nonatomic, copy) HotAction debugAction;

@property(nonatomic) ShowMode showMode;
@property(nonatomic) ShowOptions showOptions;
@property(nonatomic) RecordState recordState;

@end


@implementation UIOverlayController

- (void)dealloc
{
    DDLogDebug(@"UIOverlayController dealloc");
}

- (instancetype)initWithRootView:(UIView *)rootView
                    cameraAction:(HotAction)cameraAction
                       micAction:(HotAction)micAction
                      showAction:(HotAction)showAction
                     debugAction:(HotAction)debugAction
{
    self = [super init];
    
    if (self)
    {
        [self setRootView:rootView];
        
        [self setCameraAction:cameraAction];
        [self setMicAction:micAction];
        [self setShowAction:showAction];
        [self setDebugAction:debugAction];
        
        [self setupTouchView];
        [self setupOverlayWindow];
    }
    
    return self;
}

- (void)clean
{
    [[self hotView] removeFromSuperview];
    [[self overlayWindow] setHidden:YES];
    [self setOverlayWindow:nil];
}

- (UIView *)hotView
{
    return [self touchView];
}

- (void)setAnimator:(Animator *)animator
{
    _animator = animator;
    
    [[self overlayVC] setAnimator:animator];
}

- (void)setMode:(ShowMode)mode
{
    _showMode = mode;
    
    [[self overlayWindow] setAlpha:mode == ShowNothing? 0 : 1];
    
    [[self touchView] setShowMode:mode];
    
    [[self touchView] setProcessTouches:NO];
    
    [[self overlayVC] setShowMode:mode withAnimationCompletion:^(BOOL finish)
     {
         [self enableTouchesOnFinishAnimation:finish];
     }];
    
    [self viewWillTransitionToSize:[[self rootView] bounds].size];
}

- (void)setOptions:(ShowOptions)options
{
    _showOptions = options;
    
    [[self touchView] setShowOptions:options];
    [[self overlayVC] setShowOptions:options withAnimationCompletion:^(BOOL finish)
     {
     }];
}

- (void)setRecordState:(RecordState)state
{
    DDLogDebug(@"setRecordState");
    
    _recordState = state;
    
    [[self touchView] setRecordState:state];
    
    if (state == RecordStatePhoto)
    {
        [[self touchView] setProcessTouches:NO];
    }
    
    [[self overlayVC] setRecordState:state withAnimationCompletion:^(BOOL finish)
     {
         [self enableTouchesOnFinishAnimation:finish];
     }];
    
    [self viewWillTransitionToSize:[[self rootView] bounds].size];
}

- (void)setMicEnabled:(BOOL)micEnabled
{
    [[self overlayVC] setMicrophoneEnabled:micEnabled withAnimationCompletion:^(BOOL finish)
     { }];
}

- (void)setARKitInterruption:(BOOL)interruption
{
    [[self overlayWindow] setAlpha:interruption? 1 : 0];
}

- (void)setTrackingState:(NSString *)state sceneHasPlanes:(BOOL)hasPlanes {

    [[self overlayVC] setTrackingState:state withAnimationCompletion:^(BOOL finish) {} sceneHasPlanes:hasPlanes];
}

- (void)viewWillTransitionToSize:(CGSize)size
{
    CGRect updRect = CGRectMake(0, 0, size.width, size.height);
    
    if ([self showMode] >= ShowMulti)
    {
        if ([self showOptions] & Browser)
        {
            if (([self recordState] == RecordStateIsReady) ||
                ([self recordState] == RecordStateGoingToRecording) ||
                ([self recordState] >= RecordStatePreviewing))
            {
                updRect.origin.y = URL_BAR_HEIGHT;
            }
        }
    }
    
    [[self touchView] setCameraRect:recordFrameIn(updRect)
                            micRect:micFrameIn(updRect)
                           showRect:showFrameIn(updRect)
                          debugRect:debugFrameIn(updRect)];
}

#pragma mark Private

- (void)setupTouchView
{
    [self setTouchView:[[TouchView alloc] initWithFrame:[[self rootView] bounds]
                                           cameraAction:[self cameraAction]
                                              micAction:[self micAction]
                                             showAction:[self showAction]
                                            debugAction:[self debugAction]]];
    
    [self viewWillTransitionToSize:[[self rootView] bounds].size];
    
    [[self rootView] addSubview:[self touchView]];
    
    [[[[self touchView] topAnchor] constraintEqualToAnchor:[[self rootView] topAnchor]] setActive: YES];
    [[[[self touchView] bottomAnchor] constraintEqualToAnchor:[[self rootView] bottomAnchor]] setActive: YES];
    [[[[self touchView] leftAnchor] constraintEqualToAnchor:[[self rootView] leftAnchor]] setActive: YES];
    [[[[self touchView] rightAnchor] constraintEqualToAnchor:[[self rootView] rightAnchor]] setActive: YES];
    
    [[self touchView] setBackgroundColor:[UIColor clearColor]];
}

- (void)setupOverlayWindow
{
    UIWindow *mainWindow = [[[UIApplication sharedApplication] delegate] window];
    [self setOverlayWindow:[[UIWindow alloc] initWithFrame:[mainWindow bounds]]];
    
    [self setOverlayVC:[[OverlayViewController alloc] init]];
    [[[self overlayVC] view] setFrame:[[self overlayWindow] bounds]];
    [[self overlayWindow] setRootViewController:[self overlayVC]];
    [[self overlayWindow] setBackgroundColor:[UIColor clearColor]];
    [[self overlayWindow] setHidden:NO];
    [[self overlayWindow] setAlpha:0];
    [[self overlayWindow] setUserInteractionEnabled:NO];
    [[[self overlayVC] view] setUserInteractionEnabled:NO];
    
    dispatch_async(dispatch_get_main_queue(), ^
                   {
                       [mainWindow makeKeyWindow];
                   });
}

- (void)enableTouchesOnFinishAnimation:(BOOL)finish
{
    if (finish)
    {
        [[self touchView] setProcessTouches:YES];
    }
    else
    {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)([[self animator] animationDuration] * NSEC_PER_SEC)), dispatch_get_main_queue(), ^
                       {
                           [[self touchView] setProcessTouches:YES];
                       });
    }
}

@end

File: XRViewer/Overlay/UIOverlayController.swift
import UIKit
import XCGLogger

typealias OnSwipeDown = () -> Void
typealias OnSwipeUp = () -> Void

class UIOverlayController: NSObject {
    var animator: Animator? {
        didSet(animator) {
            overlayVC?.animator = animator
        }
    }
    var onSwipeDown: OnSwipeDown?
    var onSwipeUp: OnSwipeUp?
    private weak var rootView: UIView?
    private var touchView: TouchView?
    private var overlayWindow: UIWindow?
    private var overlayVC: OverlayViewController?
    private var debugAction: HotAction?
    private var showMode: ShowMode?
    private var showOptions: ShowOptions?

    init(rootView: UIView, debugAction: @escaping HotAction) {
        super.init()
        self.rootView = rootView
        self.debugAction = debugAction
        setupTouchView()
        setupOverlayWindow()
    }

    func clean() {
        hotView()?.removeFromSuperview()
        overlayWindow?.isHidden = true
        self.overlayWindow = nil
    }

    func viewWillTransition(to size: CGSize) {
        var updRect = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        guard let showMode = showMode else { return }
        guard let showOptions = showOptions else { return }
        
        if showMode.rawValue >= ShowMode.url.rawValue {
            if showOptions.rawValue & ShowOptions.browser.rawValue != 0 {
                updRect.origin.y = CGFloat(Constant.urlBarHeight())
            }
        }

        touchView?.setCameraRect(recordFrameIn(viewRect: updRect), micRect: micFrameIn(viewRect: updRect), showRect: showFrameIn(viewRect: updRect), debugRect: debugFrameIn(viewRect: updRect))
    }

    func hotView() -> UIView? {
        return touchView
    }

    func setMode(_ mode: ShowMode) {
        showMode = mode
        overlayWindow?.alpha = mode == .nothing ? 0 : 1
        touchView?.showMode = mode
        touchView?.setProcessTouches(false)
        overlayVC?.setShowMode(mode, withAnimationCompletion: { finish in
            self.enableTouches(onFinishAnimation: finish)
        })
        viewWillTransition(to: rootView?.bounds.size ?? CGSize.zero)
    }

    func setOptions(_ options: ShowOptions) {
        showOptions = options
        touchView?.showOptions = options
        overlayVC?.setShowOptions(options, withAnimationCompletion: { finish in
        })
    }

    func setARKitInterruption(_ interruption: Bool) {
        overlayWindow?.alpha = interruption ? 1 : 0
    }

    deinit {
        appDelegate().logger.debug("UIOverlayController dealloc")
    }

// MARK: Private

    func setupTouchView() {
        guard let rootView = rootView else { return }
        guard let debugAction = debugAction else { return }
        self.touchView = TouchView(frame: rootView.bounds, debugAction: debugAction)

        viewWillTransition(to: rootView.bounds.size)

        if let aView = touchView {
            rootView.addSubview(aView)
        }
        touchView?.backgroundColor = UIColor.clear
    }

    func setupOverlayWindow() {
        let mainWindow: UIWindow? = (UIApplication.shared.delegate?.window)!
        self.overlayWindow = UIWindow(frame: mainWindow?.bounds ?? CGRect.zero)

        self.overlayVC = OverlayViewController()
        overlayVC?.view.frame = overlayWindow?.bounds ?? CGRect.zero
        overlayWindow?.rootViewController = overlayVC
        overlayWindow?.backgroundColor = UIColor.clear
        overlayWindow?.isHidden = false
        overlayWindow?.alpha = 0
        overlayWindow?.isUserInteractionEnabled = false
        overlayVC?.view.isUserInteractionEnabled = false

        DispatchQueue.main.async(execute: {
            mainWindow?.makeKey()
        })
    }

    func enableTouches(onFinishAnimation finish: Bool) {
        if finish {
            touchView?.setProcessTouches(true)
        } else {
            guard let animationDuration = animator?.animationDuration else { return }
            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(UInt64(animationDuration) * NSEC_PER_SEC), execute: {
                self.touchView?.setProcessTouches(true)
            })
        }
    }
    
    // MARK: - UI Placement Helpers
    
    func recordFrameIn(viewRect: CGRect) -> CGRect {
        return CGRect(x: viewRect.size.width - Constant.recordSize() - Constant.recordOffsetX(), y: viewRect.origin.y + (viewRect.size.height - viewRect.origin.y) / 2 - Constant.recordSize() / 2, width: Constant.recordSize(), height: Constant.recordSize())
    }
    
    func micFrameIn(viewRect: CGRect) -> CGRect {
        return CGRect(x: viewRect.size.width - Constant.recordSize() - Constant.recordOffsetX() + (Constant.recordSize() - Constant.micSizeW()) / 2, y: viewRect.origin.y + Constant.recordOffsetY(), width: Constant.micSizeW(), height: Constant.micSizeH())
    }
    
    func debugFrameIn(viewRect: CGRect) -> CGRect {
        return CGRect(x: Constant.recordOffsetX(), y: viewRect.size.height - Constant.recordOffsetY() - Constant.micSizeH(), width: Constant.micSizeW(), height: Constant.micSizeH())
    }
    
    func showFrameIn(viewRect: CGRect) -> CGRect {
        return CGRect(x: viewRect.size.width - Constant.recordOffsetX() - Constant.micSizeW(), y: viewRect.size.height - Constant.recordOffsetY() - Constant.micSizeH(), width: Constant.micSizeW(), height: Constant.micSizeH())
    }
}
